lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,471 @@
+                    )
+                    for entry in created_csr
+                ]
+            )
+            else True
+        )
+
+    def _should_request_new_certificate(self) -> bool:
+        """Determines if we should request a new certificate from the tls relation."""
+        if not self.charm.model.relations.get(TLS_CERTIFICATE_ACCESS_RELATION_NAME):
+            return False
+        csr_exists = (existing_csr := self._pull_tls_file_from_workload(File.CSR))
+        if not csr_exists:
+            return True
+        csr_is_for_current_certificate = csr_matches_certificate(
+            existing_csr, self._pull_tls_file_from_workload(File.CERT)
+        )
+        certificate_request_pending = self._new_certificate_request_pending()
+        new_certificate_available = self.tls_access._find_certificate_in_relation_data(
+            existing_csr
+        )
+        certificate_expiring = False
+        # TODO: we probably should prefer either a CRL or a method in the library for this
+        if csr_is_for_current_certificate and not certificate_expiring:
+            return False
+        if certificate_request_pending:
+            return False
+        if new_certificate_available:
+            return False
+        return True
+
+    def _send_new_certificate_request_to_provider(
+        self, old_csr: Optional[str], ingress_address: str
+    ) -> None:
+        """This function creates and sends a new certificate signing request to the provider.
+
+        Args:
+            old_csr: Optional value that is used to decide wether to send a renewal or new request.
+            ingress_address: string that is the subject of the certificate.
+        """
+        if not (private_key := self._pull_tls_file_from_workload(File.KEY)):
+            private_key = generate_private_key().decode()
+            self._push_tls_file_to_workload(File.KEY, private_key)
+
+        new_csr = generate_csr(
+            private_key=private_key.encode(),
+            subject=ingress_address,
+            sans_ip=[ingress_address],
+            sans_dns=[socket.getfqdn()],
+        )
+        self._push_tls_file_to_workload(File.CSR, new_csr.decode())
+        if old_csr:
+            self.tls_access.request_certificate_renewal(old_csr.encode(), new_csr)
+        else:
+            self.tls_access.request_certificate_creation(new_csr)
+
+    def _get_tls_file_path_in_charm(self, file: File) -> str:
+        """Returns the requested file's location in the charm (not in the workload).
+
+        This path would typically be: /var/lib/juju/storage/certs/0/{file}.pem
+
+        Args:
+            file: a File object that determines which file path to return
+        Returns:
+            str: path
+        Raises:
+            VaultCertsError: If the CA certificate is not found
+        """
+        storage = self.charm.model.storages
+        if "certs" not in storage:
+            raise VaultCertsError()
+        if len(storage["certs"]) == 0:
```

Comment:
PEP-8 would recommend this:
```suggestion
        if not storage["certs"]:
```

I do not feel strongly about it.

Reply:
This function was copied from elsewhere. Changed.